_source/_posts/2021-11-08-k8s-api-server-oidc.md

```diff
@@ -0,0 +1,562 @@
+
+{% include toc.md %}
+
+# Why secure the API server?
+
+The API server has multiple layers of security. By default, all communication with the API server uses TLS. Authentication is done using service account tokens, bearer tokens, basic authentication, a proxy, or client certificates, depending on the platform. In the case of PaaS like Amazon EKS, AKS, and GKE, it can also be done using custom authentication mechanisms. Once a request is authenticated, the API server can use one of several authorization mechanisms, like Attribute-based access control (ABAC) or Role-based access control (RBAC), to control access to resources. And finally, there are also admission control modules that can be configured to control resource modifications.
+
+Since the API server is the only part of the Kubernetes cluster that a client can access, it is essential to secure the API server. Unauthorized access to the API server can lead to hijacking of the entire cluster, which also lets an attacker access all secrets and data accessible to the services running on it. Hence, configuring users and roles properly is a must to secure the cluster, especially i9n organizations where more than one user can access the cluster.
+
+# Why OpenID Connect?
+
+> [OpenID Connect](https://developer.okta.com/docs/concepts/oauth-openid/) is an authentication standard built on top of OAuth 2.0. It adds an additional token called an ID token. OpenID Connect also standardizes areas that OAuth 2.0 leaves up to choices, such as scopes, endpoint discovery, and dynamic registration of clients.
+
+While the default authentication mechanism of the Kubernetes API server might be enough for simple use cases, where only a handful of people manages the cluster, it does not scale for bigger organizations. It's definitely not the most secure way as well. This is because Kubernetes does not handle user management and expects this to be done outside. This is where OpenID Connect (OIDC) comes in. OIDC can manage users and groups, which works very well with the Kubernetes RBAC to provide very granular control of who can access what inside a cluster.
+
+{% img blog/k8s-api-server-oidc/kube-login-oidc.jpg alt:"K8s OIDC flow" width:"800" %}{: .center-image }
+
+Having an OIDC integration also means you can use the same OIDC provider used to do SSO in your existing infrastructure to access your Kubernetes cluster as well, like Okta or Keycloak, for example.
+
+# Use Okta as an OIDC provider to secure the API server
+
+> Okta, Inc. is an identity and access management company providing cloud software that helps companies manage and secure user authentication into applications and for developers to build identity controls into applications, websites, web services, and devices. Okta is a certified OpenID Connect provider.
```

Comment:
```suggestion
> Okta provides cloud software that helps companies manage and secure user authentication into applications and enables developers to build identity controls into applications, websites, web services, and devices. Okta is a certified OpenID Connect provider.
```
so it's not verbatim from the wikipedia company description, while keeping the focus on the technology.